### PR TITLE
Flink compiler refactor

### PR DESCRIPTION
**Pull request by arkadius**
## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

**Comment by wrzontek** [comment]
i think you mean "refactor", not "reactor" :D

**Review by wrzontek — APPROVED** [review]
just some nitpicks and questions

File: engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/SyncInterpretationFunction.scala [hunk 1]
```diff
@@ -18,17 +18,17 @@ import scala.concurrent.{Await, ExecutionContext, Future}
...
+    val compilerDataForClassloader: ClassLoader => FlinkProcessCompilerData,
     val node: SplittedNode[_ <: NodeData],
     validationContext: ValidationContext,
     useIOMonad: Boolean
 ) extends RichFlatMapFunction[Context, InterpretationResult]
     with ProcessPartFunction {
 
   private lazy implicit val ec: ExecutionContext = SynchronousExecutionContext.ctx
-  private lazy val compiledNode                  = compiledProcessWithDeps.compileSubPart(node, validationContext)
+  private lazy val compiledNode                  = compilerData.compileSubPart(node, validationContext)
```
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/flink/executor/src/main/scala/pl/touk/nussknacker/engine/process/registrar/SyncInterpretationFunction.scala:29]: I think this import only obfuscates what is used from `compilerData`, I'll just remove it.

File: interpreter/src/main/scala/pl/touk/nussknacker/engine/Interpreter.scala [hunk 2]
```diff
@@ -79,7 +79,7 @@ private class InterpreterInternal[F[_]](
           case Some(FragmentOutputVarDefinition(varName, fields)) =>
             // TODO simplify
             val withModifiedVariable = createOrUpdateVariable(ctx, varName, fields)
-            parentContext.withVariable(varName, withModifiedVariable(varName))
+            parentContext.withVariable(varName, withModifiedVariable[Any](varName))
```
[the remark this answers is not in the document]
Comment (arkadius) [reply interpreter/src/main/scala/pl/touk/nussknacker/engine/Interpreter.scala]: When you provide methods that are downcasting to some type given in the generic parameter, you should ensure that nobody provides `Nothing` by mistake. It is easy to do this kind of mistake, because in scala, when you don't provide a generic parameter and it can't be inferred from other parameters, scala implicitly put a `Nothing` into it. We have such a solution in `Typed.fromDetailedType`. I checked in how many places we forgot to specify the type. But after that I realize that `Context` is a part of `component-api` and I shouldn't change it now. For now, I left only a comment next to `Context` that we should change it.
TL;DR
I reverted this.
